### PR TITLE
Tozpatron 161 users roles

### PR DESCRIPTION
**Pull request by maras93**
- user roles entities
- super admin with basic authentication
- user operations are only available for super admin

### CODE REVIEW

File: src/main/java/com/intive/patronage/toz/users/UserService.java [hunk 1]
```diff
@@ -11,53 +12,58 @@
...
 
-    private void throwNotFoundExceptionIfNotExists(final UUID id) {
+    private void throwNotFoundExceptionIfIdNotExists(final UUID id) {
         if (!userRepository.exists(id)) {
             throw new NotFoundException(USER);
         }
     }
 
-    void throwNotFoundExceptionIfNotExists(final String email) {
+    public User findOneByEmail(String email) {
```
Comment (macieg-b) [inline src/main/java/com/intive/patronage/toz/users/UserService.java]: Is it correct name for method which can only throw exception? Why did you change name from "throw*Exception"?

File: src/main/java/com/intive/patronage/toz/users/UsersController.java [hunk 2]
```diff
@@ -45,32 +35,43 @@
...
+    public ResponseEntity<UserView> createUser(@Valid @RequestBody UserView userView) {
+        final User createdUser = userService.createWithPassword(convertToModel(userView), userView.getPassword());
+        final UserView createdUserView = convertToView(createdUser);
         final URI baseLocation = ServletUriComponentsBuilder.fromCurrentRequest()
                 .build().toUri();
-        final String userLocationString = String.format("%s%s", baseLocation, createdUser.getId());
+        final String userLocationString = String.format("%s/%s", baseLocation, createdUser.getId());
         final URI location = UriComponentsBuilder.fromUriString(userLocationString).build().toUri();
-        return ResponseEntity.created(location).body(createdUser);
+        return ResponseEntity.created(location).body(createdUserView);
```
Comment (macieg-b) [inline src/main/java/com/intive/patronage/toz/users/UsersController.java]: Do you return hash of password after create user?

Comment (macieg-b) [inline src/main/java/com/intive/patronage/toz/users/UsersController.java]: I think after convert model to view it's password digest not password. I'm not sure if it's secure to show it.

File: src/main/java/com/intive/patronage/toz/users/model/db/User.java [hunk 3]
```diff
@@ -1,58 +1,44 @@
...
+import java.util.Set;
 
+@Getter
+@Setter
 @Entity
 public class User extends Identifiable {
-    private String email;
-    private String password;
-    private String forename;
+    private String name;
```
Comment (macieg-b) [inline src/main/java/com/intive/patronage/toz/users/model/db/User.java]: If name is unique value (like email or login) you have to create setter or something which can save it as lower case string (as @mlotysz said).

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 4]
```diff
@@ -159,17 +159,43 @@ databaseChangeLog:
                       - primaryKey: true
                       - nullable: false
                 - column:
-                    name: email
+                    name: name
```
Comment (macieg-b) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: Should not it be unique value?

File: src/main/java/com/intive/patronage/toz/config/SecurityConfig.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                .and().headers().frameOptions().disable();
+    }
+
+    @Override
+    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
+        auth.authenticationProvider(superAdminAuthenticationProvider);
+    }
+
+    @Component
+    private static class SuperAdminAuthenticationProvider implements AuthenticationProvider {
```
Comment (LukaszPiskadlo) [inline src/main/java/com/intive/patronage/toz/config/SecurityConfig.java]: Wouldn't it be better to put this class in separate file?

File: src/main/java/com/intive/patronage/toz/pet/PetsController.java [hunk 6]
```diff
@@ -49,7 +49,7 @@
     @ApiOperation(value = "Get all pets", responseContainer = "List")
     @GetMapping
     public List<PetView> getAllPets(@RequestParam(value = "admin", required = false) boolean isAdmin) {
-        if (isAdmin) {
+        if (isAdmin) { // TODO
```
Comment (LukaszPiskadlo) [inline src/main/java/com/intive/patronage/toz/pet/PetsController.java]: Maybe add description what should be done in this TODO.

File: src/main/java/com/intive/patronage/toz/users/UserService.java (hunk 1 above)
Comment (maras93) [inline src/main/java/com/intive/patronage/toz/users/UserService.java]: Precisely check that exception will be thrown only when email not exist, otherwise `User` will be returned.

File: src/main/java/com/intive/patronage/toz/users/UsersController.java (hunk 2 above)
Comment (maras93) [inline src/main/java/com/intive/patronage/toz/users/UsersController.java]: No information about password is returned. `UserView` has property `password` but only with access to write. It's unavailable to read by other methods.

File: src/main/java/com/intive/patronage/toz/users/model/db/User.java (hunk 3 above)
Comment (maras93) [inline src/main/java/com/intive/patronage/toz/users/model/db/User.java]: I think that `name` shouldn't be uniqe because there  could be many users with same name e.g `Jochny` :). However this change should be apply to `email`.

File: src/main/java/com/intive/patronage/toz/config/SecurityConfig.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        @Autowired
+        public SuperAdminAuthenticationProvider(UserService userService, PasswordEncoder passwordEncoder) {
+            this.userService = userService;
+            this.passwordEncoder = passwordEncoder;
+        }
+
+        @Override
+        public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+            final String name = authentication.getName();
+            final User user = userService.findOneByName(name);
```
Comment (Jakub-Izbicki) [inline src/main/java/com/intive/patronage/toz/config/SecurityConfig.java]: Why do you look for super admin by name? Is super admin's name unique?

File: src/main/java/com/intive/patronage/toz/config/SecurityConfig.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                Collection<? extends GrantedAuthority> authorities =
+                        Collections.singleton(new SimpleGrantedAuthority(superAdminRole));
+                return new UsernamePasswordAuthenticationToken(name, authentication.getCredentials(), authorities);
+            }
+            throw new BadCredentialsException(BAD_CREDENTIALS_MESSAGE);
+        }
+
+        private void checkPasswordMatching(Authentication authentication, User user) {
+            String password = (String) authentication.getCredentials();
+            boolean isPasswordMatch = passwordEncoder.matches(password, user.getPasswordHash());
```
Comment (Jakub-Izbicki) [inline src/main/java/com/intive/patronage/toz/config/SecurityConfig.java]: Doesn't it need a `TODO` comment? E.g. `implement getPasswordHash() when hashes in db` So team knows whats going on.

Comment (maras93) [inline src/main/java/com/intive/patronage/toz/config/SecurityConfig.java]: Hashing is already implemented.

Comment (Jakub-Izbicki) [inline src/main/java/com/intive/patronage/toz/config/SecurityConfig.java]: Ugh, right. My intellij must have messed something up while changing working branch, and it was showing me that some methods couldnt be resolved. :) Sorry for that

File: src/main/java/com/intive/patronage/toz/users/model/db/User.java (hunk 3 above)
Comment (macieg-b) [inline src/main/java/com/intive/patronage/toz/users/model/db/User.java]: Sorry, I didn't notice that we have "email". What i wrote is proper for "email" field.

File: src/main/java/com/intive/patronage/toz/users/UsersController.java (hunk 2 above)
Comment (macieg-b) [inline src/main/java/com/intive/patronage/toz/users/UsersController.java]: Ok

File: src/test/java/com/intive/patronage/toz/tokens/TokenControllerTest.java [hunk 9]
```diff
@@ -31,25 +30,23 @@
     @Mock
     private TokensService tokensService;
     private MockMvc mockMvc;
-
-    private UserView userView;
+    private UserCredentials userCredentials;
```
Comment (Jakub-Izbicki) [inline src/test/java/com/intive/patronage/toz/tokens/TokenControllerTest.java]: Maybe `UserCredentials` could be renamed to `UserCredentialsView` to keep consistency with the rest of the project?

File: src/main/java/com/intive/patronage/toz/tokens/JwtFactory.java [hunk 10]
```diff
@@ -31,10 +33,18 @@ String generateToken(User user) {
...
                 .claim(EMAIL_CLAIM_NAME, user.getEmail())
-                .claim(SCOPES_CLAIM_NAME, Collections.singleton(user.getRole()))
+                .claim(SCOPES_CLAIM_NAME, getRolesFromUser(user))
                 .setIssuedAt(new Date(Instant.now().toEpochMilli()))
                 .setExpiration(new Date(Instant.now().plus(expirationTime, ChronoUnit.MINUTES).toEpochMilli()))
                 .signWith(SignatureAlgorithm.HS512, TextCodec.BASE64.decode(secret))
                 .compact();
     }
+
+    private static List<User.Role> getRolesFromUser(User user) {
```
Comment (LukaszPiskadlo) [inline src/main/java/com/intive/patronage/toz/tokens/JwtFactory.java]: You could use streams for this, but it's ok anyway.

File: src/main/java/com/intive/patronage/toz/users/UsersController.java [hunk 11]
```diff
@@ -1,41 +1,33 @@
...
 import org.springframework.web.util.UriComponentsBuilder;
+import springfox.documentation.annotations.ApiIgnore;
 
 import javax.validation.Valid;
 import java.net.URI;
 import java.util.List;
 import java.util.UUID;
 
 
+@ApiIgnore
```
Comment (LukaszPiskadlo) [inline src/main/java/com/intive/patronage/toz/users/UsersController.java]: Why remove this resource from documentation?

File: src/main/resources/application.properties [hunk 12]
```diff
@@ -22,8 +22,12 @@ server.external.url.context=/
 timezoneOffset=Z
 
 # security
+security.basic.enabled=true
```
Comment (LukaszPiskadlo) [inline src/main/resources/application.properties]: This is no longer needed.

File: src/main/java/com/intive/patronage/toz/users/UsersController.java (hunk 11 above)
Comment (maras93) [inline src/main/java/com/intive/patronage/toz/users/UsersController.java]: Because all operations on `User` resource should be available only for admins.

File: src/main/java/com/intive/patronage/toz/tokens/JwtFactory.java (hunk 10 above)
Comment (maras93) [inline src/main/java/com/intive/patronage/toz/tokens/JwtFactory.java]: I decided that stream would be less readable.

File: src/main/java/com/intive/patronage/toz/users/UsersController.java (hunk 11 above)
Comment (LukaszPiskadlo) [inline src/main/java/com/intive/patronage/toz/users/UsersController.java]: But this doesn't mean it shouldn't be in documentation. We are documenting whole system.

Comment (Jakub-Izbicki) [inline src/main/java/com/intive/patronage/toz/users/UsersController.java]: But then how will the clients know how to work with out api? `Users` resource will be of course available only for admins/super admin, because this will be secured with `@PreAuthorize()` etc. Task: [toz-163](https://tracker.intive.com/jira/browse/TOZPATRON-163)

Comment (maras93) [inline src/main/java/com/intive/patronage/toz/users/UsersController.java]: See `TOZPATRON-161` description, there is requirement that this API should be hidden.

Comment (LukaszPiskadlo) [inline src/main/java/com/intive/patronage/toz/users/UsersController.java]: If it's specified by task, then ok, let's hide this.

Comment (Jakub-Izbicki) [inline src/main/java/com/intive/patronage/toz/users/UsersController.java]: If I remember right, the thing that was supposed to be hidden was not the `users` whole api, but only `super admin` ability to locally create new users.
